addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/internal/discovery/ZWayBridgeDiscoveryService.java

```diff
@@ -0,0 +1,186 @@
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+import org.apache.commons.net.util.SubnetUtils;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.zway.ZWayBindingConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ZWayBridgeDiscoveryService} is responsible for device discovery.
+ *
+ * @author Patrick Hecker - Initial contribution
+ */
+public class ZWayBridgeDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    private final static int SEARCH_TIME = 240;
+    private final static int INITIAL_DELAY = 15;
+    private final static int SCAN_INTERVAL = 3600;
+
+    private ZWayDeviceScan mZWayServerScanningRunnable;
+    private ScheduledFuture<?> mZWayServerScanningJob;
+
+    public ZWayBridgeDiscoveryService() {
+        super(ZWayBindingConstants.SUPPORTED_DEVICE_THING_TYPES_UIDS, SEARCH_TIME);
+        logger.debug("Initializing ZWayBridgeDiscoveryService");
+
+        mZWayServerScanningRunnable = new ZWayDeviceScan();
+        activate(null);
+    }
+
+    private void scan() {
+        logger.debug("Starting scan for Z-Way Server");
+
+        ValidateIPV4 validator = new ValidateIPV4();
+
+        try {
+            Enumeration<NetworkInterface> enumNetworkInterface = NetworkInterface.getNetworkInterfaces();
+            while (enumNetworkInterface.hasMoreElements()) {
+                NetworkInterface networkInterface = enumNetworkInterface.nextElement();
+                if (networkInterface.isUp() && !networkInterface.isVirtual() && !networkInterface.isLoopback()) {
+                    for (InterfaceAddress address : networkInterface.getInterfaceAddresses()) {
+                        if (validator.isValidIPV4(address.getAddress().getHostAddress())) {
+                            String ipAddress = address.getAddress().getHostAddress();
+                            Short prefix = address.getNetworkPrefixLength();
+
+                            logger.debug("Scan IP address for Z-Way Server: {}", ipAddress);
+
+                            // Search on localhost first
+                            scheduler.execute(new ZWayServerScan(ipAddress));
+
+                            String subnet = ipAddress + "/" + prefix;
+                            SubnetUtils utils = new SubnetUtils(subnet);
+                            String[] addresses = utils.getInfo().getAllAddresses();
+
+                            for (String addressInSubnet : addresses) {
+                                scheduler.execute(new ZWayServerScan(addressInSubnet));
```

Comment:
Woah, this for in a for loop can be scheduling a huge number of jobs... Imagine you are in a 255.255.0.0 subnet and might have several network interfaces.
It would really make sense that the Z-Way server broadcasts its existence via mDNS or UPnP...